src/libmongoc/src/mongoc/mongoc-bulkwrite.c

```diff
@@ -1438,12 +1445,36 @@ _bulkwritereturn_apply_result (mongoc_bulkwritereturn_t *self,
    return true;
 }
 
+BSON_EXPORT (void)
+mongoc_bulkwrite_set_client (mongoc_bulkwrite_t *self, mongoc_client_t *client)
+{
+   BSON_ASSERT_PARAM (self);
+   BSON_ASSERT_PARAM (client);
+
+   if (self->session) {
+      BSON_ASSERT (self->session->client == client);
+   }
+
+   /* NOP if the client is not changing; otherwise, assign it and increment and
+    * fetch its operation_id. */
+   if (self->client == client) {
+      return;
+   }
+
+   self->client = client;
+   self->operation_id = ++client->cluster.operation_id;
+}
+
 BSON_EXPORT (void)
 mongoc_bulkwrite_set_session (mongoc_bulkwrite_t *self, mongoc_client_session_t *session)
 {
    BSON_ASSERT_PARAM (self);
    BSON_OPTIONAL_PARAM (session);
 
+   if (self->client) {
```

Comment:
```suggestion
   if (self->client && session) {
```

Unlike `mongoc_bulk_operation_set_client_session`, `mongoc_bulkwrite_set_session` does not require `session` be non-NULL (uses `BSON_OPTIONAL_PARAM` instead of `BSON_ASSERT_PARAM`).